spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/Tokens.java

```diff
@@ -0,0 +1,307 @@
+		return (nextToken.getKind() == tokenKind);
+	}
+
+	/**
+	 * Return {@code true} if the indicated token matches the current token
+	 * position.
+	 *
+	 * @param desiredTokenKind token to match
+	 * @return true if the current token kind matches the provided token kind
+	 */
+	protected boolean peek(TokenKind desiredTokenKind) {
+		return peek(desiredTokenKind, false);
+	}
+
+	/**
+	 * Return {@code true} if the indicated token matches the current token
+	 * position.
+	 *
+	 * @param desiredTokenKind token to match
+	 * @param consumeIfMatched if {@code true}, advance the current token position
+	 * @return true if the current token kind matches the provided token kind
+	 */
+	private boolean peek(TokenKind desiredTokenKind, boolean consumeIfMatched) {
+		if (!hasNext()) {
+			return false;
+		}
+		Token t = peek();
+		if (t.kind == desiredTokenKind) {
+			if (consumeIfMatched) {
+				position++;
+			}
+			return true;
+		}
+		else {
+			return false;
+		}
+	}
+
+	/**
+	 * Return the next {@link Token} and advance the current token position.
+	 *
+	 * @return next {@code Token}
+	 */
+	protected Token next() {
+		if (!hasNext()) {
+			raiseException(expression.length(), JobDSLMessage.OOD);
+		}
+		return tokenStream.get(position++);
+	}
+
+	/**
+	 * Consume the next token if it matches the indicated token kind;
+	 * otherwise throw {@link CheckPointedParseException}.
+	 *
+	 * @param expectedKind the expected token kind
+	 * @return the next token
+	 * @throws CheckPointedParseException if the next token does not match
+	 * the expected token kind
+	 */
+	protected Token eat(TokenKind expectedKind) {
+		Token t = next();
+		if (t == null) {
+			raiseException(expression.length(), JobDSLMessage.OOD);
+		}
+		if (t.kind != expectedKind) {
+			raiseException(t.startpos, JobDSLMessage.NOT_EXPECTED_TOKEN,
+					expectedKind.toString().toLowerCase(),
+					t.getKind().toString().toLowerCase() + (t.data == null ? "" : "(" + t.data + ")"));
+		}
+		return t;
+	}
+
+	/**
+	 * Consume the next token if it matches the desired token kind and return true; otherwise
+	 * return false.
+	 *
+	 * @param desiredKind the desired token to eat
+	 * @return true if the token was consumed, otherwise false
+	 */
+	protected boolean maybeEat(TokenKind desiredKind) {
+		if (peek(desiredKind)) {
+			next();
+			return true;
+		}
+		else {
+			return false;
+		}
+	}
+
+
+	/**
+	 * Return {@code true} if the first character of the token at the current position
+	 * is the same as the last character of the token at the previous position.
+	 *
+	 * @return true if the first character of the current token matches the last
+	 * character of the previous token
+	 */
+	protected boolean isNextAdjacent() {
+		if (!hasNext()) {
+			return false;
+		}
+
+		Token last = tokenStream.get(position - 1);
+		Token next = tokenStream.get(position);
+		return next.startpos == last.endpos;
+	}
+
+	/**
+	 * Indicate that the current token has been successfully processed.
```

Comment:
The whole "checkpoint" thing in StreamParser is about recording the last good, parseable position at the semantic level (i.e. not just token boundaries).
So for example, we should not checkpoint after just `<a & b` (missing `>`)